flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/TableEnvironmentTest.scala

```diff
@@ -18,231 +18,574 @@
+        util.addTable[JTuple3[Int, String, Double]]('a),
+        Seq("a" -> INT))
+    }
   }
 
   @Test
-  def testGetFieldInfoRowNames(): Unit = {
-    val fieldInfo = tEnv.getFieldInfo(
-      rowType,
-      Array(
-        UnresolvedFieldReference("name1"),
-        UnresolvedFieldReference("name2"),
-        UnresolvedFieldReference("name3")
-      ))
-
-    fieldInfo._1.zip(Array("name1", "name2", "name3")).foreach(x => assertEquals(x._2, x._1))
-    fieldInfo._2.zip(Array(0, 1, 2)).foreach(x => assertEquals(x._2, x._1))
+  def testStreamProjectWithAddingTimeAttributesByPosition(): Unit = {
+    val util = streamTestUtil()
+
+    // case class
+    util.verifySchema(
+      util.addTable[CClass]('a, 'b, 'c , 'proctime.proctime),
+      Seq("a" -> INT, "b" -> STRING, "c" -> DOUBLE, "proctime" -> PROCTIME))
+
+    util.verifySchema(
+      util.addTable[CClass]('a, 'b, 'c, 'rowtime.rowtime),
+      Seq("a" -> INT, "b" -> STRING, "c" -> DOUBLE, "rowtime" -> ROWTIME))
+
+    util.verifySchema(
+      util.addTable[CClass]('a, 'b, 'c, 'rowtime.rowtime, 'proctime.proctime),
+      Seq("a" -> INT, "b" -> STRING, "c" -> DOUBLE, "rowtime" -> ROWTIME, "proctime" -> PROCTIME))
+
+    // row
+    util.verifySchema(
+      util.addTable('a, 'b, 'c, 'proctime.proctime)(TEST_ROW),
+      Seq("a" -> INT, "b" -> STRING, "c" -> DOUBLE, "proctime" -> PROCTIME))
+
+    util.verifySchema(
+      util.addTable('a, 'b, 'c, 'rowtime.rowtime)(TEST_ROW),
+      Seq("a" -> INT, "b" -> STRING, "c" -> DOUBLE, "rowtime" -> ROWTIME))
+
+    util.verifySchema(
+      util.addTable('a, 'b, 'c, 'rowtime.rowtime, 'proctime.proctime)(TEST_ROW),
+      Seq("a" -> INT, "b" -> STRING, "c" -> DOUBLE, "rowtime" -> ROWTIME, "proctime" -> PROCTIME))
+
+    // tuple
+    util.verifySchema(
+      util.addTable[JTuple3[Int, String, Double]]('a, 'b, 'c, 'proctime.proctime),
+      Seq("a" -> INT, "b" -> STRING, "c" -> DOUBLE, "proctime" -> PROCTIME))
+
+    util.verifySchema(
+      util.addTable[JTuple3[Int, String, Double]]('a, 'b, 'c, 'rowtime.rowtime),
+      Seq("a" -> INT, "b" -> STRING, "c" -> DOUBLE, "rowtime" -> ROWTIME))
+
+    util.verifySchema(
+      util.addTable[JTuple3[Int, String, Double]]('a, 'b, 'c, 'rowtime.rowtime, 'proctime.proctime),
+      Seq("a" -> INT, "b" -> STRING, "c" -> DOUBLE, "rowtime" -> ROWTIME, "proctime" -> PROCTIME))
   }
 
   @Test
-  def testGetFieldInfoTuple(): Unit = {
-    val fieldInfo = tEnv.getFieldInfo(tupleType)
-
-    fieldInfo._1.zip(Array("f0", "f1", "f2")).foreach(x => assertEquals(x._2, x._1))
-    fieldInfo._2.zip(Array(0, 1, 2)).foreach(x => assertEquals(x._2, x._1))
+  def testStreamAliasWithReplacingTimeAttributesByPosition(): Unit = {
+    val util = streamTestUtil()
+
+    // case class
+    util.verifySchema(
+      util.addTable[CClassWithTime]('a, 'b.rowtime, 'c),
+      Seq("a" -> INT, "b" -> ROWTIME, "c" -> STRING))
+
+    util.verifySchema(
+      util.addTable[CClassWithTime]('a, ('b as 'new).rowtime, 'c),
```

Comment:
which field does `'b` reference? We could do the same with `'new.rowtime`, right?

Reply:
Yes we could. I just tested the case. But it does make much sense.